### PR TITLE
hdfs vdb based example

### PR DESCRIPTION
**Pull request by aditya300899**
When I run the application I get this error: Caused by: org.teiid.deployers.VirtualDatabaseException: TEIID40095 TEIID31080 accounts.stock_price validation error: TEIID30357 file.getTextFiles does not exist.
I am not able to figure this out yet! What do you think @rareddy @shawkins ?

### CODE REVIEW

File: samples/hdfs/src/main/resources/customer-vdb.ddl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+CREATE VIEW stock_price (
+	symbol string(255) NOT NULL AUTO_INCREMENT,
+	price double NOT NULL,
+	PRIMARY KEY(symbol)
+)
+AS
+SELECT
+tt.symbol, tt.price
+FROM (EXEC file.getTextFiles('/user/aditya/stock.txt')) AS f,
```
Comment (rareddy) [inline samples/hdfs/src/main/resources/customer-vdb.ddl:35]: this should be accounts.getTextFiles

File: samples/hdfs/src/main/java/org/teiid/spring/example/DataSources.java [hunk 2]
```diff
@@ -15,26 +15,18 @@
...
-    @Autowired
-    public HdfsConnectionFactory mysite(@Qualifier("mysite-config") HdfsConfiguration config) {
-        return new HdfsConnectionFactory(config);
-    }
-
-    @Bean(name="mysite-config")
-    @ConfigurationProperties("spring.teiid.data.hdfs.mysite")
-    public HdfsConfiguration mysiteConfig() {
-        return new HdfsConfiguration();
+    @ConfigurationProperties(prefix = "spring.datasource.mydb")
```
Comment (rareddy) [inline samples/hdfs/src/main/java/org/teiid/spring/example/DataSources.java:27]: The previous one is correct, you need to build the HDFSConnectionFactory here, and mydb properties correspond to the configuring where the HDFS is

File: samples/hdfs/src/main/resources/application.properties [hunk 3]
```diff
@@ -16,7 +16,9 @@
 
 spring.main.allow-bean-definition-overriding=true
 
-spring.teiid.data.hdfs.mysite.fsUri = hdfs://localhost:9000
+spring.datasource.mydb.fsUri = hdfs://localhost:9000
```
Comment (rareddy) [inline samples/hdfs/src/main/resources/application.properties:19]: use as "spring.teiid.data.hdfs.mydb.xxx" here where xxx are individual properties

File: samples/hdfs/src/main/java/org/teiid/spring/example/DataSources.java (hunk 2 above)
Comment (rareddy) [inline samples/hdfs/src/main/java/org/teiid/spring/example/DataSources.java:27]: use as "spring.teiid.data.hdfs.mydb" , typically "spring.datasource" reserved for jdbc type properties

**Comment by aditya300899** [comment]
the connection is all good but this pops up on the logs:
java.lang.IllegalStateException: Failed to execute CommandLineRunner
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:787) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:768) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.teiid.spring.example.Application.main(Application.java:35) ~[classes/:na]
Caused by: org.springframework.jdbc.UncategorizedSQLException: StatementCallback; uncategorized SQLException for SQL [SELECT *  FROM stock_price]; SQL state [38000]; error code [0]; org.teiid.core.TeiidException; nested exception is org.teiid.jdbc.TeiidSQLException: org.teiid.core.TeiidException
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:89) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1443) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:388) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:452) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:462) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.queryForList(JdbcTemplate.java:490) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.teiid.spring.example.Application.run(Application.java:40) ~[classes/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:784) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	... 5 common frames omitted
Caused by: org.teiid.jdbc.TeiidSQLException: org.teiid.core.TeiidException
	at org.teiid.jdbc.TeiidSQLException.create(TeiidSQLException.java:131) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.TeiidSQLException.create(TeiidSQLException.java:67) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.postReceiveResults(StatementImpl.java:786) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.access$100(StatementImpl.java:62) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl$2.onCompletion(StatementImpl.java:584) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.client.util.ResultsFuture.addCompletionListener(ResultsFuture.java:144) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:580) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:423) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeQuery(StatementImpl.java:354) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.springframework.jdbc.core.JdbcTemplate$1QueryStatementCallback.doInStatement(JdbcTemplate.java:439) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:376) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	... 10 common frames omitted
Caused by: org.teiid.core.TeiidException: null
	at org.teiid.client.ResultsMessage.setException(ResultsMessage.java:189) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.sendError(RequestWorkItem.java:1172) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.close(RequestWorkItem.java:607) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.process(RequestWorkItem.java:386) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.AbstractWorkItem.run(AbstractWorkItem.java:43) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.run(RequestWorkItem.java:284) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPCore.executeRequest(DQPCore.java:361) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPCore.executeRequest(DQPCore.java:276) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.teiid.logging.LogManager$LoggingProxy.invoke(LogManager.java:119) ~[teiid-api-15.0.0-20200624.204946-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.transport.SessionCheckingProxy.invoke(SessionCheckingProxy.java:60) ~[teiid-runtime-15.0.0-20200624.205033-10.jar:15.0.0-SNAPSHOT]
	at com.sun.proxy.$Proxy51.executeRequest(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.teiid.transport.LocalServerConnection$1$1.call(LocalServerConnection.java:212) ~[teiid-runtime-15.0.0-20200624.205033-10.jar:15.0.0-SNAPSHOT]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at org.teiid.dqp.internal.process.DQPWorkContext.runInContext(DQPWorkContext.java:281) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPWorkContext.runInContext(DQPWorkContext.java:265) ~[teiid-engine-15.0.0-20200624.204951-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.transport.LocalServerConnection$1.invoke(LocalServerConnection.java:210) ~[teiid-runtime-15.0.0-20200624.205033-10.jar:15.0.0-SNAPSHOT]
	at com.sun.proxy.$Proxy51.executeRequest(Unknown Source) ~[na:na]
	at org.teiid.jdbc.StatementImpl.execute(StatementImpl.java:748) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:578) ~[teiid-client-15.0.0-20200624.204943-10.jar:15.0.0-SNAPSHOT]
	... 14 common frames omitted
Caused by: java.lang.StackOverflowError: null
	at org.teiid.file.VirtualFile$1.getStorageMode(VirtualFile.java:48) ~[file-api-15.0.0-20200624.205028-10.jar:15.0.0-SNAPSHOT]

This is probably due to I think the SQL query but I don't see that it is wrong in some way. Is it that we query it in a different way?

**Comment by shawkins** [comment]
> The above getStorageMode is called a lot many times recursively.

That is a bug.  The method is mistakenly calling itself rather than the same method on the VirtualFile.  I'll correct that.

**Comment by shawkins** [comment]
Corrected with https://github.com/teiid/teiid/commit/c5bf2829a3e65111206096e113b9010150771cd1

**Comment by aditya300899** [comment]
Thanks for the fix. This works well!
I have coded the examples for both hdfs and s3. With s3 it's good!

With hdfs, there's a problem that the filesystem closes between the application and thus throws an IOException. The fix that I could come up with was to remove the close method and it worked. Is it a good fix? 
If not I'd see what else can be done.

**Comment by shawkins** [comment]
> The fix that I could come up with was to remove the close method and it worked. Is it a good fix?

Which close method specifically are you talking about?  HdfsConnection.close?  And what exception and when are you seeing it?

In general it's the job of the areFilesUsableAfterClose to tell the engine that any returned lobs can't be used after the connection is closed, so the engine will hold the execution open until the statement is closed.

**Comment by aditya300899** [comment]
> Which close method specifically are you talking about? HdfsConnection.close?

Yes.

> And what exception and when are you seeing it?

I see the following stack trace on running the application.java


java.lang.IllegalStateException: Failed to execute CommandLineRunner
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:787) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:768) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.teiid.spring.example.Application.main(Application.java:35) ~[classes/:na]
Caused by: org.springframework.jdbc.UncategorizedSQLException: StatementCallback; uncategorized SQLException for SQL [SELECT *  FROM stock_price]; SQL state [50000]; error code [30180]; TEIID30180 java.sql.SQLException: Filesystem closed; nested exception is org.teiid.jdbc.TeiidSQLException: TEIID30180 java.sql.SQLException: Filesystem closed
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:89) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1443) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:388) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:452) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:462) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.queryForList(JdbcTemplate.java:490) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.teiid.spring.example.Application.run(Application.java:40) ~[classes/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:784) ~[spring-boot-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	... 5 common frames omitted
Caused by: org.teiid.jdbc.TeiidSQLException: TEIID30180 java.sql.SQLException: Filesystem closed
	at org.teiid.jdbc.TeiidSQLException.create(TeiidSQLException.java:131) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.TeiidSQLException.create(TeiidSQLException.java:67) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.postReceiveResults(StatementImpl.java:786) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.access$100(StatementImpl.java:62) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl$2.onCompletion(StatementImpl.java:584) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.client.util.ResultsFuture.addCompletionListener(ResultsFuture.java:144) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:580) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:423) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeQuery(StatementImpl.java:354) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.springframework.jdbc.core.JdbcTemplate$1QueryStatementCallback.doInStatement(JdbcTemplate.java:439) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:376) ~[spring-jdbc-5.2.5.RELEASE.jar:5.2.5.RELEASE]
	... 10 common frames omitted
Caused by: org.teiid.core.TeiidProcessingException: TEIID30180 java.sql.SQLException: Filesystem closed
	at org.teiid.query.processor.relational.TextTableNode.initReader(TextTableNode.java:467) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.TextTableNode.nextBatchDirect(TextTableNode.java:209) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.RelationalNode.nextBatch(RelationalNode.java:277) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.BatchIterator.finalRow(BatchIterator.java:65) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.common.buffer.AbstractTupleSource.getCurrentTuple(AbstractTupleSource.java:66) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.BatchIterator.getCurrentTuple(BatchIterator.java:80) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.common.buffer.AbstractTupleSource.hasNext(AbstractTupleSource.java:88) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.NestedTableJoinStrategy.process(NestedTableJoinStrategy.java:93) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.JoinNode.nextBatchDirectInternal(JoinNode.java:260) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.JoinNode.nextBatchDirect(JoinNode.java:195) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.RelationalNode.nextBatch(RelationalNode.java:277) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.ProjectNode.nextBatchDirect(ProjectNode.java:146) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.RelationalNode.nextBatch(RelationalNode.java:277) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.RelationalPlan.nextBatch(RelationalPlan.java:139) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.QueryProcessor.nextBatchDirect(QueryProcessor.java:147) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.QueryProcessor.nextBatch(QueryProcessor.java:110) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.BatchCollector.collectTuples(BatchCollector.java:160) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.BatchCollector.collectTuples(BatchCollector.java:142) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.processMore(RequestWorkItem.java:491) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.process(RequestWorkItem.java:361) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.AbstractWorkItem.run(AbstractWorkItem.java:43) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.RequestWorkItem.run(RequestWorkItem.java:284) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPCore.executeRequest(DQPCore.java:361) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPCore.executeRequest(DQPCore.java:276) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.teiid.logging.LogManager$LoggingProxy.invoke(LogManager.java:119) ~[teiid-api-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.transport.SessionCheckingProxy.invoke(SessionCheckingProxy.java:60) ~[teiid-runtime-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at com.sun.proxy.$Proxy51.executeRequest(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.teiid.transport.LocalServerConnection$1$1.call(LocalServerConnection.java:212) ~[teiid-runtime-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at org.teiid.dqp.internal.process.DQPWorkContext.runInContext(DQPWorkContext.java:281) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.dqp.internal.process.DQPWorkContext.runInContext(DQPWorkContext.java:265) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.transport.LocalServerConnection$1.invoke(LocalServerConnection.java:210) ~[teiid-runtime-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at com.sun.proxy.$Proxy51.executeRequest(Unknown Source) ~[na:na]
	at org.teiid.jdbc.StatementImpl.execute(StatementImpl.java:748) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.jdbc.StatementImpl.executeSql(StatementImpl.java:578) ~[teiid-client-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	... 14 common frames omitted
Caused by: java.sql.SQLException: Filesystem closed
	at org.teiid.core.types.BaseLob.getBinaryStream(BaseLob.java:104) ~[teiid-common-core-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.core.types.BaseLob.getCharacterStream(BaseLob.java:97) ~[teiid-common-core-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.core.types.BaseClobType.getCharacterStream(BaseClobType.java:69) ~[teiid-common-core-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.query.processor.relational.TextTableNode.initReader(TextTableNode.java:460) ~[teiid-engine-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	... 56 common frames omitted
Caused by: java.io.IOException: Filesystem closed
	at org.apache.hadoop.hdfs.DFSClient.checkOpen(DFSClient.java:474) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.hdfs.DFSClient.open(DFSClient.java:1002) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:320) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:316) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.open(DistributedFileSystem.java:328) ~[hadoop-client-api-3.1.3.jar:na]
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:899) ~[hadoop-client-api-3.1.3.jar:na]
	at org.teiid.hdfs.HdfsVirtualFile.openInputStream(HdfsVirtualFile.java:45) ~[file-hdfs-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.file.VirtualFile$1.getInputStream(VirtualFile.java:43) ~[file-api-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	at org.teiid.core.types.BaseLob.getBinaryStream(BaseLob.java:102) ~[teiid-common-core-15.0.0-SNAPSHOT.jar:15.0.0-SNAPSHOT]
	... 59 common frames omitted

**Comment by shawkins** [comment]
> I see the following stack trace on running the application.java

The expectation is that the connection should be held open.  Can you confirm is this is happening with just a single connection?

From what I could see with the hdfs filesystem support FileSystem.get would return a cached instance - not an instance per connection.  Since it was unclear if that was thread-safe, I attempted to disable the caching - but used: configuration.set("s.hdfs.impl.disable.cache", "true"); //$NON-NLS-1$ //$NON-NLS-2$
which should have been:
configuration.set("fs.hdfs.impl.disable.cache", "true"); //$NON-NLS-1$ //$NON-NLS-2$

That means that if two connections are used, then close on one of them will likely close both until that is corrected.

**Comment by shawkins** [comment]
I can see this is happening in teiid-spring-boot due to separate connections being used for the metadata/capabilities initialization and the execution and because the explicit close detection doesn't work if the lobs are already typed as ClobType or BlobType (which the file translator does).  I'll make a correction there as well.

So in reviewing this more I believe our usage of the filesystem is thread-safe - the underlying DFSClient is appropriately thread-safe for our operations.  To not rely on the caching mechansim, https://github.com/teiid/teiid/pull/1306 adds explicit sharing of a single filesystem instance.  I'll update teiid spring boot

**Pull request merged**